File: src/test/java/aserralle/akka/stream/kcl/javadsl/Examples.java
/*
 * Copyright (C) 2018 Albert Serrallé
 */

package aserralle.akka.stream.kcl.javadsl;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;
import aserralle.akka.stream.kcl.CommittableRecord;
import aserralle.akka.stream.kcl.KinesisWorkerCheckpointSettings;
import aserralle.akka.stream.kcl.KinesisWorkerSourceSettings;
import scala.concurrent.duration.FiniteDuration;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient;
import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
import software.amazon.awssdk.services.kinesis.KinesisAsyncClient;
import software.amazon.kinesis.common.ConfigsBuilder;
import software.amazon.kinesis.coordinator.Scheduler;
import software.amazon.kinesis.processor.ShardRecordProcessorFactory;
import software.amazon.kinesis.retrieval.KinesisClientRecord;
import software.amazon.kinesis.retrieval.RetrievalConfig;
import software.amazon.kinesis.retrieval.polling.SimpleRecordsFetcherFactory;
import software.amazon.kinesis.retrieval.polling.SynchronousBlockingRetrievalFactory;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class Examples {

    //#init-client
    Region region = Region.AWS_GLOBAL;
    KinesisAsyncClient kinesisClient = KinesisAsyncClient.builder().region(region).build();
    DynamoDbAsyncClient dynamoClient = DynamoDbAsyncClient.builder().region(region).build();
    CloudWatchAsyncClient cloudWatchClient = CloudWatchAsyncClient.builder().region(region).build();
    //#init-client

    //#init-system
    final ActorSystem system = ActorSystem.create();
    final ActorMaterializer materializer = ActorMaterializer.create(system);
    //#init-system

    //#worker-settings
    final KinesisWorkerSource.WorkerBuilder workerBuilder = new KinesisWorkerSource.WorkerBuilder() {
        @Override
        public Scheduler build(ShardRecordProcessorFactory recordProcessorFactory) {

            String streamName = "myStreamName";

            ConfigsBuilder configsBuilder =
                    new ConfigsBuilder(
                            streamName,
                            "myApp",
                            kinesisClient,
                            dynamoClient,
                            cloudWatchClient,
                            "workerId",
                            recordProcessorFactory);

            //#Fan-out retrievalConfig - will incur additional AWS costs
            RetrievalConfig fanoutRetrievalConfig =
                    configsBuilder.retrievalConfig();
            //#Fan-out retrievalConfig

            //#Non-fan-out retrievalConfig i.e. equivalent of KCL 1 client
            RetrievalConfig retrievalConfig =
                    configsBuilder.retrievalConfig().retrievalFactory(
                            new SynchronousBlockingRetrievalFactory(
                                    streamName,
                                    kinesisClient,
                                    new SimpleRecordsFetcherFactory(),
                                    1000));
            //#Non-fan-out retrievalConfig

            return new Scheduler(
                    configsBuilder.checkpointConfig(),
                    configsBuilder.coordinatorConfig(),
                    configsBuilder.leaseManagementConfig(),
                    configsBuilder.lifecycleConfig(),
                    configsBuilder.metricsConfig(),
                    configsBuilder.processorConfig(),
                    retrievalConfig);
        }
    };

    final KinesisWorkerSourceSettings workerSettings = KinesisWorkerSourceSettings.create(
            1000,
            FiniteDuration.apply(1L, TimeUnit.SECONDS), FiniteDuration.apply(1L, TimeUnit.MINUTES));
    //#worker-settings

    //#worker-source
    final Executor workerExecutor = Executors.newFixedThreadPool(100);
    final Source<CommittableRecord, Scheduler> workerSource =
        KinesisWorkerSource.create(workerBuilder, workerSettings, workerExecutor );
    //#worker-source

    //#checkpoint
    final KinesisWorkerCheckpointSettings checkpointSettings =
            KinesisWorkerCheckpointSettings.create(1000, FiniteDuration.apply(30L, TimeUnit.SECONDS));
    final Flow<CommittableRecord, KinesisClientRecord, NotUsed> checkpointFlow =
        KinesisWorkerSource.checkpointRecordsFlow(checkpointSettings);
    //#checkpoint

}

File: src/test/scala/aserralle/akka/stream/kcl/scaladsl/Examples.scala
/*
 * Copyright (C) 2018 Albert Serrallé
 */

package aserralle.akka.stream.kcl.scaladsl

import java.util.UUID
import java.util.concurrent.Executors

import akka.actor.ActorSystem
import akka.stream.scaladsl.Sink
import akka.stream.{ActorMaterializer, Materializer}
import aserralle.akka.stream.kcl.{KinesisWorkerCheckpointSettings, KinesisWorkerSourceSettings}
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient
import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
import software.amazon.awssdk.services.kinesis.KinesisAsyncClient
import software.amazon.kinesis.common.ConfigsBuilder
import software.amazon.kinesis.coordinator.Scheduler
import software.amazon.kinesis.processor.ShardRecordProcessorFactory
import software.amazon.kinesis.retrieval.polling.{SimpleRecordsFetcherFactory, SynchronousBlockingRetrievalFactory}

import scala.concurrent.ExecutionContext
import scala.concurrent.duration._
import scala.language.postfixOps

object Examples {

  //#init-system
  implicit val system: ActorSystem = ActorSystem()
  implicit val materializer: Materializer = ActorMaterializer()
  //#init-system

  //#init-clients
  val region: Region = Region.EU_WEST_1
  val kinesisClient: KinesisAsyncClient = KinesisAsyncClient.builder.region(region).build
  val dynamoClient: DynamoDbAsyncClient = DynamoDbAsyncClient.builder.region(region).build
  val cloudWatchClient: CloudWatchAsyncClient = CloudWatchAsyncClient.builder.region(region).build
  //#init-clients

  //#worker-settings
  val workerSourceSettings = KinesisWorkerSourceSettings(
    bufferSize = 1000,
    terminateStreamGracePeriod = 1 minute,
    backpressureTimeout = 1 minute)

  val builder: ShardRecordProcessorFactory => Scheduler = recordProcessorFactory => {

    val streamName = "myStreamName"

    val configsBuilder = new ConfigsBuilder(
      streamName,
      "myApp",
      kinesisClient,
      dynamoClient,
      cloudWatchClient,
      s"${
        import scala.sys.process._
        "hostname".!!.trim()
      }:${UUID.randomUUID()}",
      recordProcessorFactory)

    //#Fan-out retrievalConfig - will incur additional AWS costs
    val fanoutRetrievalConfig = configsBuilder.retrievalConfig
    //#Fan-out retrievalConfig

    //#Non-fan-out retrievalConfig i.e. equivalent of KCL 1 client
    val retrievalConfig =
      configsBuilder.retrievalConfig
          .retrievalFactory(
            new SynchronousBlockingRetrievalFactory(
              streamName,
              kinesisClient,
              new SimpleRecordsFetcherFactory,
              1000))
    //#Non-fan-out retrievalConfig


    new Scheduler(
      configsBuilder.checkpointConfig,
      configsBuilder.coordinatorConfig,
      configsBuilder.leaseManagementConfig,
      configsBuilder.lifecycleConfig,
      configsBuilder.metricsConfig,
      configsBuilder.processorConfig,
      configsBuilder.retrievalConfig)
  }
  //#worker-settings

  //#worker-source
  implicit val executor =
    ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(1000))

  KinesisWorkerSource(builder, workerSourceSettings).to(Sink.ignore)
  //#worker-source

  //#checkpoint
  val checkpointSettings = KinesisWorkerCheckpointSettings(100, 30 seconds)
  KinesisWorkerSource(builder, workerSourceSettings)
    .via(KinesisWorkerSource.checkpointRecordsFlow(checkpointSettings))
    .to(Sink.ignore)
  KinesisWorkerSource(builder, workerSourceSettings).to(
    KinesisWorkerSource.checkpointRecordsSink(checkpointSettings))
  //#checkpoint

}
